### ISSUE TITLE
Upload Error: socket hang up

### ISSUE THREAD

**Issue by Barry127** [qgustavor/mega#31]
Uploading a file using storage.upload throws an Error: socket hang up.

```
fs.createReadStream('./myFile').pipe(s.upload('myFile'));
```

```
Error: socket hang up
    at createHangUpError (_http_client.js:330:15)
    at Socket.socketOnEnd (_http_client.js:423:23)
    at Socket.emit (events.js:164:20)
    at endReadableNT (_stream_readable.js:1054:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

I've got the same issue on the original mega repo. I cannot figure it out.

**Comment by qgustavor** [comment]
From [Stack Overflow](https://stackoverflow.com/questions/16995184/nodejs-what-does-socket-hang-up-actually-mean) seems MEGA is closing the connection. Currently uploading isn't being tested automatically. For now I have some test scripts which uploads files to a test account I created. I ran those: upload isn't working.

From what I remember uploading code is the same from the original library (I just converted it to use modern javascript, but the logic is the same). Probably MEGA changed how uploading works. I will check [official](https://github.com/meganz/webclient/) and [megatools](https://github.com/megous/megatools/) repositories to check what changed.

**Comment by Barry127** [comment]
Thanks for all the research. I've been puzzling with it for two days before posting.
I've got no idea what happened, my knowledge about the subject is far less then yours as I conclude by reading your answer.

**Comment by qgustavor** [comment]
Strange, seems that now it's working: https://directme.ga/view?F!J0VEwBxI!!MkcXwSJJ&c=a.png

Check the date, I just uploaded it: https://mega.nz/#F!J0VEwBxI!DeontologiaPROSOPOPEIA (I use two random words as keys for testing, it's something gfycat-like)

I just don't know yet if it's working because the changes I just did or if MEGA changed something.

**Comment by Barry127** [comment]
It's not working for me:

```
fs.createReadStream('./package.json').pipe(s.upload('package.json'))
Stream { ... }
> Error: socket hang up
    at createHangUpError (_http_client.js:330:15)
    at Socket.socketOnEnd (_http_client.js:423:23)
    at Socket.emit (events.js:164:20)
    at endReadableNT (_stream_readable.js:1054:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

Any suggestions?

**Comment by qgustavor** [comment]
Now it's not working for me too. If you need it urgently just download megatools and use `require('child_process')`: seems it will take some time to fix this bug.

I'm planning to record TCP streams (using pcap) then I will try to make the library behave the most possible similar way as the web client does.

I tried testing uploading using the browser version, but I just discovered that uploading in browser never worked in the first place: I never tested it, I find insecure entering MEGA credentials in a third party library.

**Comment by qgustavor** [comment]
The server just is being stricter: it's requiring `Connection: Keep-Alive` to be set. Maybe that's because they expect clients to do parallel uploading, in this case reusing connections is good for them. I just find bad that they just made their server close the connection without returning any error (which would help debugging).

Two files I just uploaded: [test 1](https://directme.ga/view?F!cg0x3BLI!ProsopopeiaCAPITALISTA!9tEXCTrC&c=i.jpg) and [test 2](https://directme.ga/view?F!p5sHnJAS!!w0NjSb5J&c=a.jpg).

You can overwrite request module defaults using the following code  
([request.defaults](https://github.com/request/request#requestdefaultsoptions) returns a wrapper, so you need to overwrite it):

```javascript
storage.api.requestModule = storage.api.requestModule.defaults({
  forever: true,
  headers: {connection: 'keep-alive'}
})
```

I will release a new version that fixes uploading soon, so you don't need to use the above code. Well, it's still useful: I use it to configure user-agent and proxy settings (although I need to document it in wiki).

**Comment by qgustavor** [comment]
Fixed in 3daa01b

**Comment by SahilAggarwal2004** [comment]
Its been a while since this issue was opened first, but I am encountering this issue and I don't know how to fix it :(
